File: backend/app/channel_agent/discovery_policy.py
from __future__ import annotations

from dataclasses import dataclass
import re
from typing import Any


class DiscoveryPolicyError(ValueError):
    pass


@dataclass(frozen=True)
class DiscoveryPolicy:
    enabled: bool = False
    interval_minutes: int = 360
    max_queries_per_run: int = 3
    max_results_per_query: int = 10
    min_view_count: int = 1000
    region_code: str = "US"

    @classmethod
    def from_content_mix(cls, value: object) -> DiscoveryPolicy:
        if not isinstance(value, dict):
            raise DiscoveryPolicyError("content mix policy must be an object")

        nested_value = value.get("youtube_discovery", {})
        if not isinstance(nested_value, dict):
            raise DiscoveryPolicyError("youtube discovery policy must be an object")

        enabled = _exact_bool(nested_value, "enabled", cls.enabled)
        interval_minutes = _bounded_int(
            nested_value,
            "interval_minutes",
            cls.interval_minutes,
            minimum=60,
            maximum=1440,
        )
        max_queries_per_run = _bounded_int(
            nested_value,
            "max_queries_per_run",
            cls.max_queries_per_run,
            minimum=1,
            maximum=5,
        )
        max_results_per_query = _bounded_int(
            nested_value,
            "max_results_per_query",
            cls.max_results_per_query,
            minimum=1,
            maximum=25,
        )
        min_view_count = _bounded_int(
            nested_value,
            "min_view_count",
            cls.min_view_count,
            minimum=0,
            maximum=1_000_000_000,
        )
        region_value = (
            nested_value["region_code"]
            if "region_code" in nested_value
            else value.get("region_code", cls.region_code)
        )
        region_code = _region_code(region_value)

        return cls(
            enabled=enabled,
            interval_minutes=interval_minutes,
            max_queries_per_run=max_queries_per_run,
            max_results_per_query=max_results_per_query,
            min_view_count=min_view_count,
            region_code=region_code,
        )


def _exact_bool(value: dict[str, Any], field: str, default: bool) -> bool:
    parsed = value.get(field, default)
    if type(parsed) is not bool:
        raise DiscoveryPolicyError(f"{field} must be a boolean")
    return parsed


def _bounded_int(
    value: dict[str, Any],
    field: str,
    default: int,
    *,
    minimum: int,
    maximum: int,
) -> int:
    parsed = value.get(field, default)
    if type(parsed) is not int:
        raise DiscoveryPolicyError(f"{field} must be an integer")
    if not minimum <= parsed <= maximum:
        raise DiscoveryPolicyError(
            f"{field} must be between {minimum} and {maximum}"
        )
    return parsed


def _region_code(value: object) -> str:
    if type(value) is not str or re.fullmatch(r"[A-Z]{2}", value) is None:
        raise DiscoveryPolicyError("region_code must be two uppercase ASCII letters")
    return value

File: internal/channelops/discovery_policy.go
package channelops

import (
	"encoding/json"
	"fmt"
	"strconv"
)

const discoverySourceYouTubeSearch = "youtube_search"

type DiscoveryPolicy struct {
	Enabled            bool
	IntervalMinutes    int
	MaxQueriesPerRun   int
	MaxResultsPerQuery int
	MinViewCount       int
	RegionCode         string
}

var defaultDiscoveryPolicy = DiscoveryPolicy{
	Enabled:            false,
	IntervalMinutes:    360,
	MaxQueriesPerRun:   3,
	MaxResultsPerQuery: 10,
	MinViewCount:       1000,
	RegionCode:         "US",
}

// DiscoveryPolicyFromContentMix mirrors the Python policy parser for JSON-decoded content mixes.
func DiscoveryPolicyFromContentMix(contentMix map[string]any) (DiscoveryPolicy, error) {
	if contentMix == nil {
		return DiscoveryPolicy{}, fmt.Errorf("content mix policy must be an object")
	}

	nested := map[string]any{}
	if raw, exists := contentMix["youtube_discovery"]; exists {
		var ok bool
		nested, ok = raw.(map[string]any)
		if !ok || nested == nil {
			return DiscoveryPolicy{}, fmt.Errorf("youtube discovery policy must be an object")
		}
	}

	enabled, err := discoveryBool(nested, "enabled", defaultDiscoveryPolicy.Enabled)
	if err != nil {
		return DiscoveryPolicy{}, err
	}
	interval, err := discoveryInt(nested, "interval_minutes", defaultDiscoveryPolicy.IntervalMinutes, 60, 1440)
	if err != nil {
		return DiscoveryPolicy{}, err
	}
	queries, err := discoveryInt(nested, "max_queries_per_run", defaultDiscoveryPolicy.MaxQueriesPerRun, 1, 5)
	if err != nil {
		return DiscoveryPolicy{}, err
	}
	results, err := discoveryInt(nested, "max_results_per_query", defaultDiscoveryPolicy.MaxResultsPerQuery, 1, 25)
	if err != nil {
		return DiscoveryPolicy{}, err
	}
	minViews, err := discoveryInt(nested, "min_view_count", defaultDiscoveryPolicy.MinViewCount, 0, 1_000_000_000)
	if err != nil {
		return DiscoveryPolicy{}, err
	}
	region := defaultDiscoveryPolicy.RegionCode
	if raw, exists := nested["region_code"]; exists {
		region, err = discoveryRegion(raw)
		if err != nil {
			return DiscoveryPolicy{}, err
		}
	} else if raw, exists := contentMix["region_code"]; exists {
		region, err = discoveryRegion(raw)
		if err != nil {
			return DiscoveryPolicy{}, err
		}
	}

	return DiscoveryPolicy{
		Enabled:            enabled,
		IntervalMinutes:    interval,
		MaxQueriesPerRun:   queries,
		MaxResultsPerQuery: results,
		MinViewCount:       minViews,
		RegionCode:         region,
	}, nil
}

func discoveryBool(value map[string]any, field string, fallback bool) (bool, error) {
	raw, exists := value[field]
	if !exists {
		return fallback, nil
	}
	parsed, ok := raw.(bool)
	if !ok {
		return false, fmt.Errorf("%s must be a boolean", field)
	}
	return parsed, nil
}

func discoveryInt(value map[string]any, field string, fallback, minimum, maximum int) (int, error) {
	raw, exists := value[field]
	if !exists {
		return fallback, nil
	}
	number, ok := raw.(json.Number)
	if !ok || !isLexicallyIntegralJSONNumber(number.String()) {
		return 0, fmt.Errorf("%s must be an integer between %d and %d", field, minimum, maximum)
	}
	parsed, err := strconv.ParseInt(number.String(), 10, 64)
	if err != nil || parsed < int64(minimum) || parsed > int64(maximum) {
		return 0, fmt.Errorf("%s must be an integer between %d and %d", field, minimum, maximum)
	}
	return int(parsed), nil
}

func isLexicallyIntegralJSONNumber(value string) bool {
	if value == "" {
		return false
	}
	start := 0
	if value[0] == '-' {
		start = 1
	}
	if start == len(value) {
		return false
	}
	if value[start] == '0' {
		return start+1 == len(value)
	}
	if value[start] < '1' || value[start] > '9' {
		return false
	}
	for index := start + 1; index < len(value); index++ {
		if value[index] < '0' || value[index] > '9' {
			return false
		}
	}
	return true
}

func discoveryRegion(value any) (string, error) {
	region, ok := value.(string)
	if !ok || len(region) != 2 || region[0] < 'A' || region[0] > 'Z' || region[1] < 'A' || region[1] > 'Z' {
		return "", fmt.Errorf("region_code must be two uppercase ASCII letters")
	}
	return region, nil
}
